torchgeo/datamodules/utils.py

```diff
@@ -19,6 +22,81 @@ class MisconfigurationException(Exception):
     """Exception used to inform users of misuse with Lightning."""
 
 
+class AugPipe(Module):
+    """Pipeline for applying augmentations sequentially on select data keys."""
+
+    def __init__(
+        self, augs: Callable[[dict[str, Any]], dict[str, Any]], batch_size: int
+    ) -> None:
+        """Initialize a new AugPipe instance.
+
+        Args:
+            augs: Augmentations to apply.
+            batch_size: Batch size
+        """
+        super().__init__()
+        self.augs = augs
+        self.batch_size = batch_size
+
+    def forward(self, batch: dict[str, Tensor]) -> dict[str, Tensor]:
+        """Apply the augmentation.
+
+        Args:
+            batch: Input batch.
+
+        Returns:
+            Augmented batch.
+        """
+        batch_len = len(batch["image"])
+        for bs in range(batch_len):
+            batch_dict = {
+                "image": batch["image"][bs],
+                "labels": batch["labels"][bs],
+                "boxes": batch["boxes"][bs],
+            }
+
+            if "masks" in batch:
+                batch_dict["masks"] = batch["masks"][bs]
+
+            batch_dict = self.augs(batch_dict)
+
+            batch["image"][bs] = batch_dict["image"]
+            batch["labels"][bs] = batch_dict["labels"]
+            batch["boxes"][bs] = batch_dict["boxes"]
+
+            if "masks" in batch:
+                batch["masks"][bs] = batch_dict["masks"]
+
+        # Stack images
+        batch["image"] = rearrange(batch["image"], "b () c h w -> b c h w")
+
+        return batch
+
+
+def collate_fn_detection(batch: list[dict[str, Tensor]]) -> dict[str, Any]:
+    """Custom collate fn for object detection & instance segmentation.
+
+    Args:
+        batch: list of sample dicts return by dataset
+
+    Returns:
+        batch dict output
+    """
+    output: dict[str, Any] = {}
```

Comment:
Could we instead use `torchgeo.datasets.unbind_samples` and add one extra line to modify labels?

Reply:
`unbind_samples` is used to convert dict of lists into list of dicts. Here we're doing the opposite.

Reply:
Then what about concat/merge_samples? Those do the opposite.

Reply:
Do you mean `stack_samples`? I think concat and merge samples were for `GeoDatasets`.

Either way, those functions can't be used since it assumes the samples can be stacked/concatenated. This is not the case here.

Reply:
Both GeoDatasets and NonGeoDatasets use the same `sample` dict format, so all collation functions are appropriate for both. Although these tend to be more used with GeoDatasets.

I see now that your output dict has lists of Tensors instead of Tensors, which is indeed different than all other builtin collation functions. I guess this is required because each sample may have a different number of objects. Not sure how people normally handle collation for that case in PyTorch.